src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt

```diff
@@ -0,0 +1,213 @@
+package io.libp2p.pubsub.gossip
+
+import io.libp2p.core.PeerId
+import io.libp2p.core.pubsub.ValidationResult
+import io.libp2p.etc.types.cappedDouble
+import io.libp2p.etc.types.createLRUMap
+import io.libp2p.etc.types.millis
+import io.libp2p.etc.util.P2PService
+import pubsub.pb.Rpc
+import java.util.Optional
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import kotlin.math.max
+import kotlin.math.min
+import kotlin.math.pow
+
+class GossipScore(
+    val params: GossipScoreParams = GossipScoreParams(),
+    val executor: ScheduledExecutorService,
+    val curTime: () -> Long
+) {
+
+    inner class TopicScores(val topic: Topic) {
+        private val params = topicParams[topic]
+
+        var joinedMeshTime: Long = 0
+        var firstMessageDeliveries: Double by cappedDouble(
+            0.0,
+            this@GossipScore.params.decayToZero,
+            params.FirstMessageDeliveriesCap
+        )
+        var meshMessageDeliveries: Double by cappedDouble(
+            0.0,
+            this@GossipScore.params.decayToZero,
+            params.MeshMessageDeliveriesCap
+        )
+        var meshFailurePenalty: Double by cappedDouble(0.0, this@GossipScore.params.decayToZero)
+        var invalidMessages: Double by cappedDouble(0.0, this@GossipScore.params.decayToZero)
+
+        fun inMesh() = joinedMeshTime > 0
+
+        fun meshTimeNorm() = min(
+            (if (inMesh()) curTime() - joinedMeshTime else 0).toDouble() / params.TimeInMeshQuantum.toMillis(),
```

Comment:
The spec suggests that the _"The time in mesh is calculated lazily during the decay update to avoid a large number of calls to `gettimeofday`_". 

I might not be correctly understanding the lazy keyword concept in Kotlin, but it seems to me that we are not lazily updating `meshTime` during the decay update, but calling `System.currentTimeMillis()` at every `score()` call.

Reply:
The current implementation is very far from the optimal with regards to performance. `System.currentTimeMillis()` is absolutely not the bottleneck at the moment. 
I'm not sure why that is mentioned in the spec. 
May be @vyzo could comment on this?

Reply:
It's very fast in most OSes, due to VDSO.

Reply:
But then again some OSes have a very slow implementation (a syscall) -- see OpenBSD.

Reply:
It is mentioned in the spec so that there is leeway to implement it this way, if you care to avoid calling `gettimeofday` in every score invocation -- it could be called potentially for every peer within the heartbeat, hence the reason for laziness.

Reply:
Another point that is relevant, with the lazy implementation we can avoid biasing the time in mesh by order of score invocation. This _could_ be an issue if you have a slow `gettimeofday` in your system.